diff -up Linux-PAM-1.3.1/modules/pam_console/sed-static.console-build Linux-PAM-1.3.1/modules/pam_console/sed-static
--- Linux-PAM-1.3.1/modules/pam_console/sed-static.console-build	2014-01-31 14:17:53.000000000 +0100
+++ Linux-PAM-1.3.1/modules/pam_console/sed-static	2018-09-10 15:06:04.115302315 +0200
@@ -13,6 +13,7 @@ sed '
 /^YY_BUFFER_STATE yy_scan_buffer/s/^/STATIC /
 /^YY_BUFFER_STATE yy_scan_string/s/^/STATIC /
 /^void yy_switch_to_buffer/s/^/STATIC /
+/^extern int yylex/s/^extern /STATIC /
 /define YY_DECL int yylex/s/YY_DECL /YY_DECL STATIC /
 /^int yyparse/s/^/STATIC /
 /^void yyrestart/s/^/STATIC /
