diff -up Linux-PAM-1.6.1/doc/Makefile.am.noflex Linux-PAM-1.6.1/doc/Makefile.am
--- Linux-PAM-1.6.1/doc/Makefile.am.noflex	2024-04-09 18:22:59.000000000 +0200
+++ Linux-PAM-1.6.1/doc/Makefile.am	2024-04-10 11:09:39.304086982 +0200
@@ -2,7 +2,7 @@
 # Copyright (c) 2005, 2006 Thorsten Kukuk <kukuk@suse.de>
 #
 
-SUBDIRS = man specs sag adg mwg
+SUBDIRS = man sag adg mwg
 
 CLEANFILES = *~
 DISTCLEANFILES = custom-html.xsl custom-man.xsl
diff -up Linux-PAM-1.6.1/Makefile.am.noflex Linux-PAM-1.6.1/Makefile.am
--- Linux-PAM-1.6.1/Makefile.am.noflex	2024-04-10 11:09:39.304086982 +0200
+++ Linux-PAM-1.6.1/Makefile.am	2024-04-10 11:13:15.057352362 +0200
@@ -4,7 +4,7 @@
 
 AUTOMAKE_OPTIONS = 1.9 gnu dist-xz no-dist-gzip check-news
 
-SUBDIRS = libpam_internal libpam tests libpamc libpam_misc modules po conf \
+SUBDIRS = libpam_internal libpam tests libpamc libpam_misc modules po doc \
 	  xtests
 
 if HAVE_DOC
